File: leetcode/JavaScript/141 - LinkedListCycle.js
/**
 * Detects if a linked list contains a cycle using Floyd’s Cycle Detection Algorithm.
 * 
 * Approach:
 * - Uses two pointers: `slow` moves one step at a time, while `fast` moves two steps.
 * - If there is a cycle, `fast` will eventually meet `slow`, indicating a loop.
 * - If `fast` reaches the end (`null`), there is no cycle.
 * 
 * Time Complexity: O(N)
 * - In the worst case (when there is no cycle), `fast` traverses the entire list once, making it O(N).
 * - In the case of a cycle, `fast` moves at twice the speed of `slow`, meaning they meet within O(N) iterations.
 * 
 * Space Complexity: O(1)
 * - Uses only two pointers (`fast` and `slow`), so extra space usage is constant.
 * 
 */
const hasCycle = (head) => {
    // Two-pointer approach
    let fast = head;
    let slow = head;

    while (fast && fast.next) {
        fast = fast.next.next; // Move fast pointer two steps
        slow = slow.next; // Move slow pointer one step

        // If fast meets slow, cycle exists
        if (fast === slow) return true;
    }

    // No cycle detected
    return false;
};

File: Platforms(LeetCode, CF, CC)/LeetCode/141-LinkedListCycle.cpp
// Problem: Detect if a linked list has a cycle
// Given the head of a linked list, return true if there is a cycle, else false

// Time Complexity: O(N) — each node is visited at most twice
// Space Complexity: O(1) — only two pointers used

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *fast = head, *slow = head;

        while(fast && fast->next){
            fast = fast->next->next;  // move fast 2 steps
            slow = slow->next;        // move slow 1 step

            if(fast == slow)           // pointers meet → cycle detected
                return true;
        }

        return false; // no cycle
    }
};

int main() {
    // Example usage:
    ListNode* head = new ListNode(3);
    head->next = new ListNode(2);
    head->next->next = new ListNode(0);
    head->next->next->next = new ListNode(-4);
    head->next->next->next->next = head->next; // create cycle

    Solution sol;
    cout << (sol.hasCycle(head) ? "Cycle detected" : "No cycle") << endl;

    return 0;
}
